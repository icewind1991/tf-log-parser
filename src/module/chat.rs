use crate::common::{SubjectData, SubjectId};
use crate::module::EventHandler;
use crate::raw_event::{RawEvent, RawEventType};
use crate::SubjectMap;
use std::convert::Infallible;
use steamid_ng::SteamID;

struct BareChatMessage {
    pub time: u32,
    pub subject: SubjectId,
    pub message: String,
    pub chat_type: ChatType,
}

pub struct ChatMessage {
    pub time: u32,
    pub name: String,
    pub steam_id: SteamID,
    pub message: String,
    pub chat_type: ChatType,
}

impl ChatMessage {
    fn from_bare(bare: BareChatMessage, subjects: &SubjectMap) -> Self {
        let (name, steam_id) = match &subjects[bare.subject] {
            SubjectData::Player { name, steam_id, .. } => (name.clone(), steam_id.clone()),
            _ => {
                unreachable!("only player messages are added");
            }
        };
        ChatMessage {
            time: bare.time,
            name,
            steam_id,
            message: bare.message,
            chat_type: bare.chat_type,
        }
    }
}

pub enum ChatType {
    All,
    Team,
}

#[derive(Default)]
pub struct ChatHandler(Vec<BareChatMessage>);

impl EventHandler for ChatHandler {
    type Output = Vec<ChatMessage>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(ty, RawEventType::SayTeam | RawEventType::Say)
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &RawEvent,
    ) -> Result<(), Infallible> {
        if !matches!(subject, SubjectId::Player(_)) {
            return Ok(());
        }
        match event.ty {
            RawEventType::SayTeam => self.0.push(BareChatMessage {
                time,
                subject,
                message: event.params.trim_matches('"').into(),
                chat_type: ChatType::Team,
            }),
            RawEventType::Say => self.0.push(BareChatMessage {
                time,
                subject,
                message: event.params.trim_matches('"').into(),
                chat_type: ChatType::All,
            }),
            _ => {}
        }
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        self.0
            .into_iter()
            .map(|bare| ChatMessage::from_bare(bare, subjects))
            .collect()
    }
}
