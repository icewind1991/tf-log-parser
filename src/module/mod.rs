use crate::common::SubjectId;
use crate::event::GameEvent;
use crate::raw_event::RawEventType;
use crate::SubjectMap;
pub use chat::{ChatHandler, ChatMessage, ChatType};
pub use healspread::HealSpreadHandler;
pub use lobbysettings::{
    LobbySettingsError, LobbySettingsHandler, Location, Settings as LobbySettings,
};
pub use medicstats::{MedicStats, MedicStatsHandler};
use std::convert::Infallible;
use std::error::Error;
use std::fmt::Debug;
use thiserror::Error;

mod chat;
mod healspread;
mod lobbysettings;
mod medicstats;

pub trait EventHandler: Default {
    type Output;
    type Error: Error + Debug;

    fn does_handle(&self, ty: RawEventType) -> bool;

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error>;

    fn finish(self, subjects: &SubjectMap) -> Self::Output;
}

#[derive(Default)]
pub struct HandlerStack<Head, Tail> {
    head: Head,
    tail: Tail,
}

impl<Head: EventHandler, Tail: EventHandler> EventHandler for HandlerStack<Head, Tail> {
    type Output = (Head::Output, Tail::Output);
    type Error = EitherError<Head::Error, Tail::Error>;

    fn does_handle(&self, ty: RawEventType) -> bool {
        self.head.does_handle(ty) || self.tail.does_handle(ty)
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        self.head
            .handle(time, subject, event)
            .map_err(|e| EitherError::E1(e))?;
        self.tail
            .handle(time, subject, event)
            .map_err(|e| EitherError::E2(e))?;
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        (self.head.finish(subjects), self.tail.finish(subjects))
    }
}

#[derive(Debug, Error)]
pub enum EitherError<E1: Error, E2: Error> {
    #[error("{0}")]
    E1(E1),
    #[error("{0}")]
    E2(E2),
}

/// A handler that doesn't stop the parsing on failure
pub enum OptionalHandler<Handler: EventHandler> {
    Active(Handler),
    Failed(Handler::Error),
}

impl<Handler: EventHandler> Default for OptionalHandler<Handler> {
    fn default() -> Self {
        OptionalHandler::Active(Handler::default())
    }
}

impl<Handler: EventHandler> EventHandler for OptionalHandler<Handler> {
    type Output = Result<Handler::Output, Handler::Error>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        match self {
            OptionalHandler::Active(handler) => handler.does_handle(ty),
            OptionalHandler::Failed(_) => false,
        }
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        let res = if let OptionalHandler::Active(handler) = self {
            handler.handle(time, subject, event)
        } else {
            Ok(())
        };

        if let Err(e) = res {
            dbg!(&e);
            *self = OptionalHandler::Failed(e)
        }
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        match self {
            OptionalHandler::Active(handler) => Ok(handler.finish(subjects)),
            OptionalHandler::Failed(e) => Err(e),
        }
    }
}
