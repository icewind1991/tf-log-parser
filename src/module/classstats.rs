use crate::common::{Class, ClassMap, SubjectId};
use crate::event::{DamageEvent, GameEvent, RoleChangeEvent, SpawnEvent};
use crate::module::EventHandler;
use crate::raw_event::{RawEventType, RawSubject};
use crate::{EventMeta, SubjectData, SubjectMap};
use serde::Serialize;
use std::collections::BTreeMap;

#[derive(Debug, Serialize, Default, PartialEq)]
pub struct ClassStats {
    pub kills: ClassMap<u8>,
    pub deaths: ClassMap<u8>,
    pub assists: ClassMap<u8>,
    pub damage: ClassMap<u16>,
}

#[derive(Default)]
pub struct ClassStatsHandler {
    active: bool,
    data: BTreeMap<SubjectId, ClassStatData>,
}

#[derive(Default)]
pub struct ClassStatData {
    class: Class,
    deaths: ClassMap<u8>,
}

impl ClassStatsHandler {
    fn get_class(&self, subject: &RawSubject) -> Option<Class> {
        subject
            .id()
            .ok()
            .and_then(|id| self.data.get(&id))
            .map(|data| data.class)
    }

    fn data_mut(&mut self, id: SubjectId) -> &mut ClassStatData {
        self.data.entry(id).or_default()
    }
}

impl EventHandler for ClassStatsHandler {
    type GlobalOutput = ();
    type PerSubjectData = ClassStats;
    type PerSubjectOutput = ClassStats;

    fn does_handle(ty: RawEventType) -> bool {
        matches!(
            ty,
            RawEventType::Killed
                | RawEventType::KillAssist
                | RawEventType::Damage
                | RawEventType::Spawned
                | RawEventType::RoleChange
                | RawEventType::RoundWin
                | RawEventType::RoundStart
        )
    }

    fn handle(
        &mut self,
        _meta: &EventMeta,
        subject: SubjectId,
        subject_data: &mut Self::PerSubjectData,
        event: &GameEvent,
    ) {
        match event {
            GameEvent::Spawned(SpawnEvent { class: Some(class) })
            | GameEvent::RoleChange(RoleChangeEvent { class: Some(class) }) => {
                self.data_mut(subject).class = *class;
            }
            GameEvent::RoundStart => {
                self.active = true;
            }
            GameEvent::RoundWin(_) => {
                self.active = false;
            }
            GameEvent::Killed(kill) if self.active => {
                if let Ok(target) = kill.target.id() {
                    let subject_class = self.data.get(&subject).map(|data| data.class);
                    let target_data = self.data_mut(target);
                    if let Some(subject_class) = subject_class {
                        target_data.deaths[subject_class] += 1;
                    }
                    subject_data.kills[target_data.class] += 1;
                }
            }
            GameEvent::KillAssist(assist) if self.active => {
                if let Some(target_class) = self.get_class(&assist.target) {
                    subject_data.assists[target_class] += 1;
                }
            }
            GameEvent::Damage(DamageEvent {
                damage: Some(damage),
                target,
                ..
            }) if self.active && damage > &0 && damage < &1500 => {
                if let Some(target_class) = self.get_class(target) {
                    subject_data.damage[target_class] =
                        subject_data.damage[target_class].saturating_add(*damage as u16);
                }
            }
            _ => {}
        }
    }

    fn finish_global(self, _subjects: &SubjectMap) -> Self::GlobalOutput {}

    fn finish_per_subject(
        &mut self,
        subject: &SubjectData,
        mut data: Self::PerSubjectData,
    ) -> Self::PerSubjectOutput {
        data.deaths = self.data.remove(&subject.id()).unwrap_or_default().deaths;
        data
    }
}
