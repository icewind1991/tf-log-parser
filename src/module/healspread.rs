use crate::common::{SteamId3, SubjectId};
use crate::event::GameEvent;
use crate::module::EventHandler;
use crate::raw_event::RawEventType;
use crate::SubjectMap;
use std::collections::HashMap;
use std::convert::{Infallible, TryFrom};

#[derive(Default)]
pub struct HealSpreadHandler(HashMap<SteamId3, HashMap<SteamId3, u32>>);

impl EventHandler for HealSpreadHandler {
    type Output = HashMap<SteamId3, HashMap<SteamId3, u32>>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(ty, RawEventType::Healed)
    }

    fn handle(
        &mut self,
        _time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        let healer_steam_id = if let Some(steam_id) = subject.steam_id() {
            steam_id
        } else {
            return Ok(());
        };
        if let GameEvent::Healed(heal_event) = event {
            if let Ok(target_subject) = SubjectId::try_from(&heal_event.target) {
                if let Some(target_steam_id) = target_subject.steam_id() {
                    let healed = self
                        .0
                        .entry(SteamId3(healer_steam_id))
                        .or_default()
                        .entry(SteamId3(target_steam_id))
                        .or_default();
                    *healed += heal_event.amount
                }
            }
        }
        Ok(())
    }

    fn finish(self, _subjects: &SubjectMap) -> Self::Output {
        self.0
    }
}
