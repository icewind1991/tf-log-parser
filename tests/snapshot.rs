use serde::Serialize;
use std::collections::BTreeMap;
use std::fs::read_to_string;
use test_case::test_case;
use tf_log_parser::module::{ClassStats, MedicStats};
use tf_log_parser::{parse, EventHandler, LogHandler, LogHandlerPerSubjectOutput};

#[derive(Serialize)]
struct LogResult {
    global: <LogHandler as EventHandler>::GlobalOutput,
    per_player: BTreeMap<String, LogPlayerData>,
}

#[derive(Serialize)]
struct LogPlayerData {
    stats: ClassStatsRaw,
    heals: BTreeMap<String, u32>,
    medic: MedicStats,
}

impl From<LogHandlerPerSubjectOutput> for LogPlayerData {
    fn from(value: LogHandlerPerSubjectOutput) -> Self {
        LogPlayerData {
            stats: value.class_stats.into(),
            medic: value.medic_stats,
            heals: value
                .heal_spread
                .into_iter()
                .map(|(user, heals)| (user.0.steam3(), heals))
                .collect(),
        }
    }
}

#[derive(Serialize)]
struct ClassStatsRaw {
    kills: [u16; 10],
    deaths: [u16; 10],
    assists: [u16; 10],
    damage: [u16; 10],
}

impl From<ClassStats> for ClassStatsRaw {
    fn from(value: ClassStats) -> Self {
        ClassStatsRaw {
            kills: value.kills.into(),
            deaths: value.deaths.into(),
            assists: value.assists.into(),
            damage: value.damage.into(),
        }
    }
}

#[test_case("log_6s.log")]
#[test_case("log_2788889.log")]
#[test_case("log_2892242.log")]
#[test_case("log_bball.log")]
#[test_case("log_hl.log")]
fn test_parse(name: &str) {
    let content = read_to_string(&format!("test_data/{}", name)).unwrap();
    let (global, per_player) = parse(&content).unwrap();
    let log = LogResult {
        global,
        per_player: per_player
            .into_iter()
            .map(|(key, value)| (key.0.steam3(), value.into()))
            .collect(),
    };
    insta::assert_json_snapshot!(format!("parse-{}", name), log);
}
