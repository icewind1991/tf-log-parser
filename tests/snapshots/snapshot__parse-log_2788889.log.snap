---
source: tests/snapshot.rs
expression: log
---
{
  "global": {
    "chat": [
      {
        "time": 45,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "негр",
        "chat_type": "All"
      },
      {
        "time": 75,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "можно я зайду просто попрыгаю?",
        "chat_type": "All"
      },
      {
        "time": 79,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "эксперты",
        "chat_type": "All"
      },
      {
        "time": 81,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "холокост?",
        "chat_type": "All"
      },
      {
        "time": 305,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "не услышал",
        "chat_type": "Team"
      },
      {
        "time": 322,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "сразу видно что у нас большие яйца.",
        "chat_type": "All"
      },
      {
        "time": 331,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "кирил",
        "chat_type": "All"
      },
      {
        "time": 333,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "мы ваши братья",
        "chat_type": "All"
      },
      {
        "time": 337,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "мы в чистилище",
        "chat_type": "All"
      },
      {
        "time": 340,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "медик дай мне 300 в начале",
        "chat_type": "Team"
      },
      {
        "time": 345,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "я пиздец устрою",
        "chat_type": "Team"
      },
      {
        "time": 345,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "Pootis",
        "chat_type": "All"
      },
      {
        "time": 416,
        "name": "фраcталwings",
        "steam_id": 76561198192838927,
        "message": "antibite",
        "chat_type": "All"
      },
      {
        "time": 452,
        "name": "фраcталwings",
        "steam_id": 76561198192838927,
        "message": "mojno svami switch?",
        "chat_type": "All"
      },
      {
        "time": 457,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "нет",
        "chat_type": "All"
      },
      {
        "time": 471,
        "name": "Fin",
        "steam_id": 76561198090117916,
        "message": "Cоболезнуем диппи",
        "chat_type": "All"
      },
      {
        "time": 475,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "6?",
        "chat_type": "All"
      },
      {
        "time": 477,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "4?",
        "chat_type": "All"
      },
      {
        "time": 479,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "2",
        "chat_type": "All"
      },
      {
        "time": 535,
        "name": "babyrage",
        "steam_id": 76561198411024073,
        "message": "пожалуйста",
        "chat_type": "All"
      },
      {
        "time": 537,
        "name": "babyrage",
        "steam_id": 76561198411024073,
        "message": "готов вверх",
        "chat_type": "All"
      },
      {
        "time": 550,
        "name": "Fin",
        "steam_id": 76561198090117916,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 550,
        "name": "丂匚尺ㄖ千ㄒ",
        "steam_id": 76561198165363447,
        "message": "[P-REC] Собрались петушки",
        "chat_type": "Team"
      },
      {
        "time": 550,
        "name": "ᗩ",
        "steam_id": 76561198173267766,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 550,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 550,
        "name": "фраcталwings",
        "steam_id": 76561198192838927,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 550,
        "name": "pupsomode",
        "steam_id": 76561198419106750,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 584,
        "name": "ᗩ",
        "steam_id": 76561198173267766,
        "message": "мне астерикс сказал я бессмертный",
        "chat_type": "All"
      },
      {
        "time": 593,
        "name": "фраcталwings",
        "steam_id": 76561198192838927,
        "message": "LOL?",
        "chat_type": "All"
      },
      {
        "time": 595,
        "name": "little pogchamp",
        "steam_id": 76561197984934157,
        "message": "¯\\_(ツ)_/¯",
        "chat_type": "All"
      },
      {
        "time": 613,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "знач не умер он просто на респу телепортнулся",
        "chat_type": "All"
      },
      {
        "time": 805,
        "name": "Kumis",
        "steam_id": 76561198129314304,
        "message": "BLYAT",
        "chat_type": "All"
      },
      {
        "time": 812,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "че",
        "chat_type": "All"
      },
      {
        "time": 899,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "отдай",
        "chat_type": "Team"
      },
      {
        "time": 901,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "снайпера ",
        "chat_type": "Team"
      },
      {
        "time": 903,
        "name": "ONE-TYZ",
        "steam_id": 76561198821399014,
        "message": "матвей",
        "chat_type": "Team"
      },
      {
        "time": 912,
        "name": "ᗩ",
        "steam_id": 76561198173267766,
        "message": "слав убейся умоляю...",
        "chat_type": "All"
      },
      {
        "time": 988,
        "name": "丂匚尺ㄖ千ㄒ",
        "steam_id": 76561198165363447,
        "message": "!REP CLAV",
        "chat_type": "All"
      },
      {
        "time": 1044,
        "name": "Want hug",
        "steam_id": 76561198313241168,
        "message": "пупсик крутой!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1",
        "chat_type": "All"
      },
      {
        "time": 1086,
        "name": "Grigoriy",
        "steam_id": 76561198439712695,
        "message": "ФЧФЧФЧ",
        "chat_type": "All"
      }
    ]
  },
  "per_player": {
    "[U:1:129852188]": {
      "stats": {
        "kills": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          2,
          2,
          0,
          0,
          0,
          0,
          0,
          1,
          0,
          0
        ],
        "assists": [
          3,
          2,
          2,
          1,
          0,
          0,
          1,
          0,
          0,
          0
        ],
        "damage": [
          0,
          57,
          189,
          57,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {
        "[U:1:169048576]": 540,
        "[U:1:213002038]": 2636,
        "[U:1:24668429]": 1212,
        "[U:1:458841022]": 2201,
        "[U:1:479446967]": 1906
      },
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 12.7699995,
        "avg_time_to_build": 58.333332,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 7.0333333,
        "charge_count": 3,
        "drops": 0
      }
    },
    "[U:1:169048576]": {
      "stats": {
        "kills": [
          5,
          3,
          1,
          0,
          0,
          0,
          1,
          1,
          0,
          0
        ],
        "deaths": [
          5,
          1,
          2,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          3,
          1,
          0,
          0,
          0,
          0,
          1,
          0,
          0
        ],
        "damage": [
          898,
          1075,
          412,
          0,
          0,
          0,
          180,
          264,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:189526172]": {
      "stats": {
        "kills": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          3,
          5,
          0,
          3,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          1,
          1,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          255,
          101,
          0,
          52,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {
        "[U:1:205097719]": 680,
        "[U:1:232573199]": 696,
        "[U:1:374702019]": 1449,
        "[U:1:450758345]": 1531,
        "[U:1:861133286]": 610
      },
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 15.235715,
        "avg_time_to_build": 70.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 9.3,
        "charge_count": 3,
        "drops": 0
      }
    },
    "[U:1:205097719]": {
      "stats": {
        "kills": [
          1,
          2,
          0,
          1,
          1,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          6,
          4,
          0,
          1,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          4,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          585,
          747,
          0,
          358,
          285,
          0,
          93,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:213002038]": {
      "stats": {
        "kills": [
          2,
          4,
          2,
          1,
          0,
          0,
          3,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          2,
          3,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          2,
          0,
          0,
          0,
          0,
          0,
          1,
          1,
          0,
          0
        ],
        "damage": [
          647,
          909,
          418,
          120,
          44,
          0,
          450,
          187,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:232573199]": {
      "stats": {
        "kills": [
          5,
          3,
          0,
          4,
          0,
          0,
          2,
          0,
          0,
          0
        ],
        "deaths": [
          3,
          2,
          0,
          3,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          3,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          937,
          875,
          0,
          701,
          0,
          0,
          431,
          0,
          10,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:24668429]": {
      "stats": {
        "kills": [
          3,
          2,
          0,
          0,
          1,
          0,
          2,
          3,
          0,
          0
        ],
        "deaths": [
          5,
          3,
          0,
          0,
          0,
          0,
          0,
          1,
          0,
          0
        ],
        "assists": [
          0,
          0,
          0,
          0,
          0,
          0,
          3,
          0,
          0,
          0
        ],
        "damage": [
          548,
          497,
          238,
          0,
          192,
          0,
          640,
          322,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:352975440]": {
      "stats": {
        "kills": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:374702019]": {
      "stats": {
        "kills": [
          2,
          3,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          6,
          3,
          0,
          6,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          1,
          1,
          0,
          1,
          0,
          0,
          1,
          0,
          0,
          0
        ],
        "damage": [
          392,
          867,
          0,
          292,
          0,
          0,
          313,
          0,
          17,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:450758345]": {
      "stats": {
        "kills": [
          2,
          3,
          0,
          1,
          0,
          0,
          1,
          0,
          1,
          0
        ],
        "deaths": [
          5,
          4,
          0,
          1,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          1,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          959,
          748,
          0,
          404,
          0,
          0,
          329,
          0,
          127,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:458841022]": {
      "stats": {
        "kills": [
          4,
          1,
          4,
          2,
          0,
          0,
          3,
          1,
          0,
          0
        ],
        "deaths": [
          5,
          2,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          2,
          2,
          1,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          1367,
          837,
          1365,
          414,
          176,
          0,
          534,
          173,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:479446967]": {
      "stats": {
        "kills": [
          4,
          5,
          4,
          0,
          0,
          0,
          2,
          2,
          0,
          0
        ],
        "deaths": [
          2,
          2,
          0,
          0,
          0,
          0,
          0,
          1,
          0,
          0
        ],
        "assists": [
          0,
          1,
          1,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          803,
          1041,
          519,
          62,
          125,
          0,
          626,
          272,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:861133286]": {
      "stats": {
        "kills": [
          2,
          1,
          0,
          1,
          0,
          0,
          2,
          0,
          1,
          0
        ],
        "deaths": [
          3,
          3,
          0,
          1,
          2,
          0,
          0,
          0,
          2,
          0
        ],
        "assists": [
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          353,
          767,
          0,
          136,
          391,
          0,
          343,
          0,
          150,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    }
  }
}
